cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -93,47 +88,12 @@ func WithBatchSize(batchSize int) SourcegraphClientOption {
 	}
 }
 
-// WithShouldUseGRPC enables or disables the use of gRPC for communicating with Sourcegraph.
-func WithShouldUseGRPC(useGRPC bool) SourcegraphClientOption {
-	return func(c *sourcegraphClient) {
-		c.useGRPC = useGRPC
-	}
-}
-
-// WithGRPCClient sets the gRPC client to use for communicating with Sourcegraph.
-func WithGRPCClient(client proto.ZoektConfigurationServiceClient) SourcegraphClientOption {
-	return func(c *sourcegraphClient) {
-		c.grpcClient = client
-	}
-}
-
-func newSourcegraphClient(rootURL *url.URL, hostname string, opts ...SourcegraphClientOption) *sourcegraphClient {
-	httpClient := retryablehttp.NewClient()
-	httpClient.Logger = debug
-
-	// Sourcegraph might return an error message in the body if StatusCode==500. The
-	// default behavior of the go-retryablehttp restClient is to drain the body and not
-	// to propagate the error. Hence, we call ErrorPropagatedRetryPolicy instead of
-	// DefaultRetryPolicy and augment the error with the response body if possible.
-	httpClient.CheckRetry = func(ctx context.Context, resp *http.Response, err error) (bool, error) {
-		shouldRetry, checkErr := retryablehttp.ErrorPropagatedRetryPolicy(ctx, resp, err)
-
-		if resp != nil && resp.StatusCode == http.StatusInternalServerError {
-			if b, e := io.ReadAll(resp.Body); e == nil {
-				checkErr = fmt.Errorf("%w: body=%q", checkErr, string(b))
-			}
-		}
-
-		return shouldRetry, checkErr
-	}
-
+func newSourcegraphClient(rootURL *url.URL, hostname string, grpcClient proto.ZoektConfigurationServiceClient, opts ...SourcegraphClientOption) *sourcegraphClient {
```

Comment:
kinda weird we pass in the rootURL but the caller of this function needs to dial grpc and pass that client in. Can we instead make it so this function does the dialing and maybe then returns an error?

Reply:
hmm we can, but that makes testing a bit harder because you can't as easily feed in a mocked grpc client 🤔 WDYT about that?